test/fs/test_fs_readdir_ino_matches_stat_ino.c

```diff
@@ -0,0 +1,52 @@
+#include <stdio.h>
+#include <unistd.h>
+#include <assert.h>
+#include <fcntl.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <assert.h>
+#include <errno.h>
+#include <string.h>
+#include <dirent.h>
+
+void setup() {
+  int res = open("b", O_CREAT, 0777);
+  assert(res >= 0);
+  assert(close(res) == 0);
+  assert(symlink("b", "a") == 0);
+}
+
+int main() {
+  setup();
+  int dirfd = open(".", O_RDONLY);
+  assert(dirfd > 0);
+  DIR *dirp = fdopendir(dirfd);
+  assert(dirp != NULL);
+  struct stat sta, stb;
+  assert(lstat("a", &sta) == 0);
+  assert(lstat("b", &stb) == 0);
+  // Remove execute permission from directory. This prevents us from stat'ing
+  // files in the directory in the implementation of readdir which we tried to
+  // use to fix this.
+  assert(chmod(".", 0675) == 0);
+  int a_ino = -1;
+  int b_ino = -1;
+  struct dirent *ep;
+  while ((ep = readdir(dirp))) {
```

Comment:
I think you can do `while (struct dirent *ep = readdir(dirp)) {` ?

Then you don't need the extra braces since its no longer possible to accidentally write `==`.

Reply:
The `while ()` condition has to be an expression, but this is a statement. You also have to put an expression in the condition slot of a `for()` but the init and the update are allowed to be statements.